ksql-engine/src/main/java/io/confluent/ksql/util/ExecutorWithRetries.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.util;
+
+import org.apache.kafka.common.errors.RetriableException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.concurrent.Future;
+import java.util.function.Supplier;
+
+public final class ExecutorWithRetries {
+
+  private static final int NUM_RETRIES = 5;
+  private static final int RETRY_BACKOFF_MS = 500;
+  private static final Logger log = LoggerFactory.getLogger(ExecutorWithRetries.class);
+
+  private ExecutorWithRetries(){
+  }
+
+  public static <T> T execute(final Supplier<? extends Future<T>> supplier) throws Exception {
+    int retries = 0;
+    Exception lastException = null;
+    while (retries < NUM_RETRIES) {
+      try {
+        if (retries != 0) {
+          Thread.sleep(RETRY_BACKOFF_MS);
+        }
+        return supplier.get().get();
+      } catch (Exception e) {
+        if (e instanceof RetriableException
+            || e.getCause() instanceof RetriableException) {
+          retries++;
+          log.info("Retrying request due to retriable exception. Retry no: " + retries, e);
+          lastException = e;
+        } else if (e instanceof KsqlException
```

Comment:
This is starting to get a bit muddy in its implementation. This works for the two use-cases you've pulled this functionality out from. But, it's very hard-coded to those two use-cases. I'm wondering if we can't make this more generic...

The main complication seems to be around knowing when to retry or not. Should I retry only on a retrieable exception or on any exception?  So.. how about we make the call tell us?

The second thing is that we want to make the calling code _more_ readable with this change.  So we don't want the calling code having to create CompletableFutures or adding try/catch or casting.

Pulling that all together, how about something like this:

```
public final class ExecutorWithRetries {

  private static final int NUM_RETRIES = 5;
  private static final int RETRY_BACKOFF_MS = 500;
  private static final Logger log = LoggerFactory.getLogger(ExecutorWithRetries.class);

  private ExecutorWithRetries() {
  }

  public enum RetryBehaviour {
    ALWAYS,
    ON_RETRYABLE
  }

  @FunctionalInterface
  public interface Function {
    void call() throws Exception;
  }

  // I've added a Function overload so that use-cases that don't need a return value don't need to fabricate one, and the Function type allows for exception
  public static void execute(final Function function,
                             final RetryBehaviour retryBehaviour) throws Exception {
    execute(() -> {
      function.call();
      return null;
    }, retryBehaviour);
  }

  // Though not used directly at the mo, a 'Callable' version is also handy:
  public static <T> T execute(final Callable<T> supplier,
                              final RetryBehaviour retryBehaviour) throws Exception {
    return executeSync(() -> {
      final CompletableFuture<T> f = new CompletableFuture<>();
      try {
        final T result = supplier.call();
        f.complete(result);
      } catch (final Exception e) {
        f.completeExceptionally(e);
      }
      return f;
    }, retryBehaviour);
  }
 
  // Renamed to ambiguity and `retryBehaviour` added to make things explicit.
  public static <T> T executeSync(final Supplier<? extends Future<T>> supplier,
                                  final RetryBehaviour retryBehaviour) throws Exception {
    Exception lastException = null;
    for (int retries = 0; retries < NUM_RETRIES; ++retries) {
      try {
        if (retries != 0) {
          Thread.sleep(RETRY_BACKOFF_MS);
        }
        return supplier.get().get();
      } catch (final Exception e) {
        final Throwable cause = e.getCause();

        if (cause instanceof RetriableException
            || (cause instanceof Exception && retryBehaviour == RetryBehaviour.ALWAYS)) {
          log.info("Retrying request. Retry no: " + retries, e);
          lastException = e;
        } else if (cause instanceof Exception) {
          throw (Exception) cause;
        } else {
          throw new RuntimeException(cause);
        }
      }
    }
    throw lastException;
  }
}
```

Then `DropSourceCommand` becomes:

```
private void deleteTopicIfNeeded(StructuredDataSource dataSource, boolean isValidatePhase) {
    if (!isValidatePhase && deleteTopic) {
      try {
        ExecutorWithRetries.execute(
            () -> kafkaTopicClient.deleteTopics(
                Collections.singletonList(dataSource.getKsqlTopic().getKafkaTopicName())),
            ExecutorWithRetries.RetryBehaviour.ALWAYS);
      } catch (Exception e) {
        throw new RuntimeException(e);
      }
      if (dataSource.getKsqlTopic().getKsqlTopicSerDe().getSerDe()
          == DataSource.DataSourceSerDe.AVRO) {
        try {
          ExecutorWithRetries.execute(
              () -> schemaRegistryClient
                  .deleteSubject(sourceName + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX),
              ExecutorWithRetries.RetryBehaviour.ALWAYS);
        } catch (Exception e) {
          throw new RuntimeException(e);
        }
      }
    }
  }
```

And `KafkaTopicClientImpl` has bits like this:

```
ExecutorWithRetries.executeSync(
          () -> adminClient.createTopics(Collections.singleton(newTopic)).all(),
          ExecutorWithRetries.RetryBehaviour.ON_RETRYABLE);
```

What do you think? This is just hacked together and untested, but I think it gives cleaner calling code.